core/domain/classifier_services.py

```diff
@@ -0,0 +1,534 @@
+                map(self._validate_label, labels)
+                labels_list.append(labels)
+        return docs, labels_list
+
+    def _iterate_gibbs_sampling(self, iterations, doc_ids=None):
+        """
+        Runs Gibbs sampling for iterations number of times on the provided
+        docs.
+        """
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for i in xrange(iterations):
+            statez = self._run_gibbs_sampling(doc_ids)
+
+    def _add_examples(self, examples, iterations):
+        """
+        Adds examples to the internal state of the classifier, assigns random
+        initial labels to only the added docs, and runs Gibbs sampling for
+        iterations number of iterations.
+        """
+        if len(examples) == 0:
+            return
+
+        docs, labels_list = self._parse_examples(examples)
+
+        last_num_labels = self._num_labels
+        last_num_docs = self._num_docs
+        last_num_words = self._num_words
+
+        # Increments _num_labels with any new labels
+        [map(self._get_label_id, labels) for labels in labels_list]
+        self._num_docs += len(docs)
+
+        self._b_dl = numpy.concatenate(
+            (self._b_dl, numpy.zeros(
+                (last_num_docs, self._num_labels - last_num_labels),
+                dtype=int)), axis=1)
+        self._b_dl = numpy.concatenate(
+            (self._b_dl, [
+                self._get_label_vector(labels) for labels in labels_list
+            ]), axis=0)
+        self._w_dp.extend([map(self._get_word_id, doc) for doc in docs])
+        self._c_dl = numpy.concatenate(
+            (self._c_dl, numpy.zeros(
+                (last_num_docs, self._num_labels - last_num_labels),
+                dtype=int)), axis=1)
+        self._c_dl = numpy.concatenate(
+            (self._c_dl, numpy.zeros(
+                (self._num_docs - last_num_docs, self._num_labels),
+                dtype=int)), axis=0)
+        self._c_lw = numpy.concatenate(
+            (self._c_lw, numpy.zeros(
+                (last_num_labels, self._num_words - last_num_words),
+                dtype=int)), axis=1)
+        self._c_lw = numpy.concatenate(
+            (self._c_lw, numpy.zeros(
+                (self._num_labels - last_num_labels, self._num_words),
+                dtype=int)), axis=0)
+        self._c_l = numpy.concatenate(
+            (self._c_l, numpy.zeros(
+                self._num_labels - last_num_labels, dtype=int)))
+
+        for d in xrange(last_num_docs, self._num_docs):
+            self._assign_initial_labels_to_word_instances([d])
+            self._iterate_gibbs_sampling(iterations, [d])
+
+        return xrange(last_num_docs, self._num_docs)
+
+    def add_examples_for_training(self, training_examples):
+        """
+        Adds examples to the classifier with _training_iterations number of
+        iterations.
+        """
+        return self._add_examples(training_examples, self._training_iterations)
```

Comment:
Actually, mind changed. Labels can begin with underscores, and if you set label '_default' the behavior is as predicted (the doc only has label '_default'). So back to empty label lists, because None breaks a lot of stuff that assumes iterables.

And removing the no-underscore label requirement.